syntax = "proto3";

package protocir;
option java_package = "org.jacodb.impl.grpc";

import "setup.proto";

enum CIRRecordKind {
  RecordKind_Class = 0;
  RecordKind_Union = 1;
  RecordKind_Struct = 2;
}

message CIRStructType {
  string name = 1;
  CIRRecordKind kind = 2;
  bool packed = 3;
  repeated CIRTypeID members = 4;
}

message CIRType {
  CIRTypeID id = 1;
  string name = 2;
  oneof kind {
    CIRIntType int_type = 3;
    CIRSingleType single_type = 4;
    CIRDoubleType double_type = 5;
    CIRFP16Type fp16_type = 6;
    CIRBF16Type bf16_type = 7;
    CIRFP80Type fp80_type = 8;
    CIRFP128Type fp128_type = 9;
    CIRLongDoubleType long_double_type = 10;
    CIRComplexType complex_type = 11;
    CIRPointerType pointer_type = 12;
    CIRDataMemberType data_member_type = 13;
    CIRBoolType bool_type = 14;
    CIRArrayType array_type = 15;
    CIRVectorType vector_type = 16;
    CIRFuncType func_type = 17;
    CIRMethodType method_type = 18;
    CIRExceptionInfoType exception_info_type = 19;
    CIRVoidType void_type = 20;
    CIRStructType struct_type = 21;
  }
}

message CIRIntType {
  uint32 width = 1;
  bool is_signed = 2;
}

message CIRSingleType {}

message CIRDoubleType {}

message CIRFP16Type {}

message CIRBF16Type {}

message CIRFP80Type {}

message CIRFP128Type {}

message CIRLongDoubleType { CIRTypeID underlying = 1; }

message CIRComplexType { CIRTypeID element_ty = 1; }

message CIRPointerType {
  CIRTypeID pointee = 1;
  optional uint32 addr_space = 2;
}

message CIRDataMemberType {
  CIRTypeID member_ty = 1;
  CIRTypeID cls_ty = 2;
}

message CIRBoolType {}

message CIRArrayType {
  CIRTypeID elt_ty = 1;
  uint64 size = 2;
}

message CIRVectorType {
  CIRTypeID elt_ty = 1;
  uint64 size = 2;
}

message CIRFuncType {
  repeated CIRTypeID inputs = 1;
  CIRTypeID return_type = 2;
  bool var_arg = 3;
}

message CIRMethodType {
  CIRTypeID member_func_ty = 1;
  CIRTypeID cls_ty = 2;
}

message CIRExceptionInfoType {}

message CIRVoidType {}
